// @civic/common-gateway-react/dist/cjs/button/IdentityButton.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ButtonMode = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const styled_components_1 = __importDefault(require("styled-components"));
const GatewayContext_1 = require("../gateway/GatewayContext");
const IdentityButton_utils_1 = require("./IdentityButton.utils");
const Button = styled_components_1.default.button `
  width: 180px;
  background-color: #282c34;
  outline: none;
  border: 1px solid #434343;
  height: 50px;
  line-height: 20px;
  cursor: pointer;
  position: relative;
  &:disabled {
    opacity: 0.7;
    cursor: default;
  }
  &:active {
    top: 1px;
  }
  &:disabled:active {
    top: 0px;
  }
`;
const buttonTitleStyle = {
    color: 'white',
    fontSize: '16px',
    paddingLeft: '20px',
};
const buttonStyleLight = {
    background: '#fff',
    border: '1px solid #ccc',
};
const buttonTitleStyleLight = {
    color: '#666666',
};
var ButtonMode;
(function (ButtonMode) {
    ButtonMode[ButtonMode["LIGHT"] = 0] = "LIGHT";
    ButtonMode[ButtonMode["DARK"] = 1] = "DARK";
})(ButtonMode = exports.ButtonMode || (exports.ButtonMode = {}));
const IdentityButton = (_a) => {
    var { mode = ButtonMode.DARK } = _a, props = __rest(_a, ["mode"]);
    const { requestGatewayToken, gatewayStatus } = (0, GatewayContext_1.useGateway)();
    return ((0, jsx_runtime_1.jsxs)(Button, Object.assign({ className: props.className, onClick: requestGatewayToken, type: "button", disabled: (0, IdentityButton_utils_1.isDisabled)(gatewayStatus), style: mode === ButtonMode.LIGHT ? Object.assign({}, buttonStyleLight) : {} }, { children: [(0, IdentityButton_utils_1.getIcon)(gatewayStatus), (0, jsx_runtime_1.jsx)("span", Object.assign({ style: mode === ButtonMode.LIGHT ? Object.assign(Object.assign({}, buttonTitleStyle), buttonTitleStyleLight) : buttonTitleStyle }, { children: (0, IdentityButton_utils_1.getButtonText)(gatewayStatus) }))] })));
};
exports.default = IdentityButton;
