// @civic/common-gateway-react/dist/cjs/wrapper/Wrapper.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Wrapper = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const SVGClose_1 = __importDefault(require("./SVGClose"));
const SVGLoading_1 = __importDefault(require("./SVGLoading"));
const constants_1 = require("../constants");
const containerStyle = {
    background: 'rgba(0,0,0,0.5)',
    position: 'fixed',
    top: '0',
    left: '0',
    width: '100vw',
    height: '100vh',
    overflow: 'hidden',
    /* flex position the iframe container */
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
};
/* Critical that this container has no overflow scroll all scrolling will be handled by content inside the iframe */
const containerContentStyle = {
    display: 'flex',
    flexDirection: 'column',
    maxWidth: '526px',
    width: '100%',
    maxHeight: '100',
    background: '#fff',
    borderRadius: '4px',
    boxShadow: '0px 4px 20px rgba(0, 0, 0, 0.25)',
    overflow: 'hidden',
};
const containerButtonStyle = {
    top: '5px',
    right: '5px',
    padding: '10px',
    background: 'transparent',
    border: 'none',
    borderRadius: '25px',
    height: '50px',
    width: '50px',
    cursor: 'pointer',
    margin: '10px 0 0 0',
};
const containerButtonNoLogoStyle = Object.assign(Object.assign({}, containerButtonStyle), { margin: '0 0 0 0' });
const headerStyle = {
    height: '80px',
    flexGrow: 0,
    flexShrink: 0,
    flexBasis: '80px',
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    borderRadius: '4px 4px 0 0',
    borderBottom: '1px solid #ddd',
    position: 'relative',
};
const logoStyle = {
    maxHeight: '40px',
    maxWidth: '100%',
    height: 'auto',
    width: 'auto',
};
const noLogoStyle = {
    height: '16px',
};
const DefaultWrapper = ({ children = null, onClose, logo, loaded = false }) => {
    const closeButtonStyle = logo ? containerButtonStyle : containerButtonNoLogoStyle;
    return ((0, jsx_runtime_1.jsx)("div", Object.assign({ style: Object.assign(Object.assign({}, containerStyle), { position: 'fixed', flexDirection: 'column', zIndex: 999 }) }, { children: (0, jsx_runtime_1.jsxs)("div", Object.assign({ style: Object.assign(Object.assign({}, containerContentStyle), { flexDirection: 'column', position: 'relative' }) }, { children: [logo && loaded && ((0, jsx_runtime_1.jsx)("header", Object.assign({ style: Object.assign(Object.assign({}, headerStyle), { position: 'relative' }) }, { children: (0, jsx_runtime_1.jsx)("img", { style: logoStyle, src: logo, alt: "logo" }) }))), !logo && loaded && ((0, jsx_runtime_1.jsx)("div", { style: noLogoStyle })), !loaded && (0, jsx_runtime_1.jsx)(SVGLoading_1.default, {}), (0, jsx_runtime_1.jsx)("button", Object.assign({ "data-testid": constants_1.TESTID_IFRAME_CLOSE_BTN, type: "button", onClick: onClose, style: Object.assign(Object.assign({}, closeButtonStyle), { display: loaded ? 'block' : 'none', position: 'absolute' }) }, { children: (0, jsx_runtime_1.jsx)(SVGClose_1.default, {}) })), children] })) })));
};
// eslint-disable-next-line import/prefer-default-export
const Wrapper = ({ wrapper, onClose, children = null, logo, loaded = false, }) => {
    if (wrapper) {
        const WrapperComp = wrapper;
        return ((0, jsx_runtime_1.jsxs)(WrapperComp, { children: [loaded && ((0, jsx_runtime_1.jsx)("button", Object.assign({ type: "button", onClick: onClose, style: Object.assign(Object.assign({}, containerButtonStyle), { display: loaded ? 'block' : 'none', position: 'absolute' }) }, { children: (0, jsx_runtime_1.jsx)(SVGClose_1.default, {}) }))), children] }));
    }
    return ((0, jsx_runtime_1.jsx)(DefaultWrapper, Object.assign({ onClose: onClose, logo: logo, loaded: loaded }, { children: children })));
};
exports.Wrapper = Wrapper;
