// @civic/common-gateway-react/dist/esm/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GatekeeperClient = exports.ButtonMode = exports.getTokenDescription = exports.IdentityButton = exports.GatewayProvider = exports.useGateway = exports.logger = exports.ChainType = exports.CivicPassIssuanceStatus = exports.TokenIssuanceState = exports.GatewayStatus = exports.GatekeeperAPIStatus = exports.TokenState = exports.State = void 0;
var types_1 = require("./types");
Object.defineProperty(exports, "State", { enumerable: true, get: function () { return types_1.State; } });
Object.defineProperty(exports, "TokenState", { enumerable: true, get: function () { return types_1.TokenState; } });
Object.defineProperty(exports, "GatekeeperAPIStatus", { enumerable: true, get: function () { return types_1.GatekeeperAPIStatus; } });
Object.defineProperty(exports, "GatewayStatus", { enumerable: true, get: function () { return types_1.GatewayStatus; } });
Object.defineProperty(exports, "TokenIssuanceState", { enumerable: true, get: function () { return types_1.TokenIssuanceState; } });
Object.defineProperty(exports, "CivicPassIssuanceStatus", { enumerable: true, get: function () { return types_1.CivicPassIssuanceStatus; } });
Object.defineProperty(exports, "ChainType", { enumerable: true, get: function () { return types_1.ChainType; } });
var logger_1 = require("./logger");
Object.defineProperty(exports, "logger", { enumerable: true, get: function () { return __importDefault(logger_1).default; } });
var GatewayContext_1 = require("./gateway/GatewayContext");
Object.defineProperty(exports, "useGateway", { enumerable: true, get: function () { return GatewayContext_1.useGateway; } });
Object.defineProperty(exports, "GatewayProvider", { enumerable: true, get: function () { return GatewayContext_1.GatewayProvider; } });
var button_1 = require("./button");
Object.defineProperty(exports, "IdentityButton", { enumerable: true, get: function () { return button_1.IdentityButton; } });
Object.defineProperty(exports, "getTokenDescription", { enumerable: true, get: function () { return button_1.getTokenDescription; } });
Object.defineProperty(exports, "ButtonMode", { enumerable: true, get: function () { return button_1.ButtonMode; } });
__exportStar(require("./errors"), exports);
var gatekeeperClient_1 = require("./utils/gatekeeperClient");
Object.defineProperty(exports, "GatekeeperClient", { enumerable: true, get: function () { return __importDefault(gatekeeperClient_1).default; } });
